### DISCUSSION TITLE
how do i move a window to the current desktop?

### DISCUSSION THREAD

**Discussion by Lydanne** [fltk-rs/fltk-rs#1224] [Q&A] [answered]
I want to use fltk to write a screenshot tool, but I can't find a way to move the window to my current window.

```
use fltk::{prelude::*, *};

fn main() {
    let app = app::App::default();
    let mut my_window = window::Window::new(0, -100, 400, 300, "My Window");
    my_window.end();
    my_window.make_modal(true);
    my_window.set_override();
    my_window.force_position(true);
    my_window.fullscreen(true);
    my_window.show();
    my_window.set_opacity(0.5);
    my_window.make_current();
    app.run().unwrap();
}

```

**Comment by MoAlyousef** [comment, marked as answer]
Windows usually spawn in the screen you launch them from. You can check by moving the generated executable to the screen you want, then double clicking it. To programmatically force the window to appear at a specific screen, you can use the screen functions. 
```rust
let screen_coords = app::screen_xywh(0); // 0 for what’s considered the main screen

let mut my_window = window::Window:new(screen_coords.0, screen_coords.1, …);
```

**Reply by Lydanne** [reply]
![image](https://user-images.githubusercontent.com/39021696/175754644-1325290e-659a-4653-8699-2ff36b49be72.png)
I am currently on the Num2 desktop, but the window I created is on the Num1 desktop, how can I move it to Num2.

**Reply by MoAlyousef** [reply]
You can try:
```rust
let screen_coords = app::screen_xywh(1); // which should refer to the second desktop
let mut my_window = window::Window:new(screen_coords.0, screen_coords.1, 400, 300, "My Window");
my_window.set_screen_num(1); 
```

**Reply by Lydanne** [reply]
did not achieve the expected effect

```
use fltk::{prelude::*, *};

fn main() {
    let app = app::App::default();
    let screen_coords = app::screen_xywh(1); // which should refer to the second desktop
    let mut my_window =
        window::Window::new(screen_coords.0, screen_coords.1, 400, 300, "My Window");
    // my_window.show();
    my_window.set_screen_num(1);
    app.run().unwrap();
}

```

**Reply by MoAlyousef** [reply]
Is it a multi-monitor setup? Or just a multispace setup?

**Reply by Lydanne** [reply]
![image](https://user-images.githubusercontent.com/39021696/175848812-34938825-2051-44f8-b415-e664bd6bfb88.png)

I want to achieve something like this, in the current multispace.

**Reply by MoAlyousef** [reply]
In a multispace setup, fltk apps should appear in the same multispace they're launched from, no need for screen detection. However, if you have an app that's fullscreen like vs code, macos will launch the app in a different space.

**Reply by Lydanne** [reply]
Is it possible to use NSPanel to achieve this effect?

**Reply by Lydanne** [reply]
like a floating window.

**Reply by Lydanne** [reply]
By the way, there is another problem, the window I created cannot cover the menu bar.

**Reply by MoAlyousef** [reply]
As far as I know, macos doesn’t allow decorated windows to appear above the menubar. You can create a borderless fltk window thru fltk-rs. But to modify the window level, you would need to get the window’s raw_handle(), cast to NSWindow pointer and change the level using Objective-C or the objc crate.

**Reply by Lydanne** [reply]
let me try

**Reply by MoAlyousef** [reply]
It would look like:
```rust
#[macro_use]
extern crate objc;

use fltk::{enums::*, prelude::*, *};
use objc::runtime::Object;

fn main() {
    let mut win = window::Window::new(0, 0, 400, 300, None);
    win.set_border(false);
    win.set_color(Color::from_rgb(80, 80, 80));
    win.end();
    win.show();
    win.set_opacity(0.75);
    unsafe {
        // set level to NSMainMenuWindowLevel + 2
        let _obj: *mut Object = msg_send![win.raw_handle() as *mut Object, setLevel:26];
    }
    app::run().unwrap();
}
```

**Reply by Lydanne** [reply]
![image](https://user-images.githubusercontent.com/39021696/176985603-0e7d820e-49a1-4ccc-9d3c-1dceda2b8ed7.png)
Still can't cover the menu bar.

**Reply by MoAlyousef** [reply]
Try to find the value of NSMainMenuWindowLevel on your system.

**Reply by Lydanne** [reply]
![image](https://user-images.githubusercontent.com/39021696/177000390-5d5b970a-caaf-46ee-ac3d-1f53ee71f071.png)
I don't think this option works

**Reply by MoAlyousef** [reply]
AFAIK It should be NSMainMenuWindowLevel  + 2. It might not work on windows which have borders. I'm not sure.

**Reply by MoAlyousef** [reply]
![image](https://user-images.githubusercontent.com/37966791/177004801-4771b286-2d7f-4836-b169-b534362b92f2.png)

Equivalent Cocoa code (using Objective-C++):
```objc
#import <Cocoa/Cocoa.h>

int main() {
  [NSAutoreleasePool new];
  [NSApplication sharedApplication];
  [NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];
  auto window = [NSWindow alloc];
  auto screen = [NSScreen mainScreen].frame;
  [window initWithContentRect:NSMakeRect(screen.origin.x,
                                         screen.size.height - 300, 400, 300)
                    styleMask:NSWindowStyleMaskBorderless
                      backing:NSBackingStoreBuffered
                        defer:NO];
  [window setLevel:NSMainMenuWindowLevel + 2];
  [window makeKeyAndOrderFront:nil];
  window.backgroundColor = [NSColor colorWithCalibratedHue:0
                                                saturation:1
                                                brightness:0
                                                     alpha:0.5];
  [NSApp activateIgnoringOtherApps:YES];
  [NSApp run];
  return 0;
}
```

Compiled using `clang++  -std=c++11 main.mm -framework Cocoa`

**Reply by Lydanne** [reply]
Thanks a lot, this problem has been bugging me for a week.
Will fltk support this feature?

**Reply by MoAlyousef** [reply]
I've added a helper method (set_on_top) in master:
```rust
use fltk::{enums::*, prelude::*, *};

fn main() {
    let mut win = window::Window::new(0, 0, 400, 300, None);
    win.set_border(false);
    win.set_color(Color::from_rgb(50, 50, 50));
    win.end();
    win.show();
    win.set_opacity(0.75);
    win.set_on_top();
    win.resize(0, 0, 400, 300);
    app::run().unwrap();
}
```
I'll publish a new release in 5 days hopefully.

Can you try again using the objc crate, I've added a win.resize() call in the end. I think it should work:
```rust
#[macro_use]
extern crate objc;

use fltk::{enums::*, prelude::*, *};
use objc::runtime::Object;

fn main() {
    let mut win = window::Window::new(0, 0, 400, 300, None);
    win.set_border(false);
    win.set_color(Color::from_rgb(50, 50, 50));
    win.end();
    win.show();
    win.set_opacity(0.75);
    unsafe {
        // set level to NSMainMenuWindowLevel + 2
        let _obj: *mut Object = msg_send![win.raw_handle() as *mut Object, setLevel:26];
    }
    win.resize(0,0, 400, 300); // to place the window now above the main menu
    app::run().unwrap();
}
```

**Reply by Lydanne** [reply]
nice, he can work!
Can you tell me the principle.
![image](https://user-images.githubusercontent.com/39021696/177007486-0d0a2792-7db3-4f1f-9231-77737a91d322.png)

**Reply by MoAlyousef** [reply]
Sure.
We acquire the window's native handle, using win.raw_handle() which returns an opaque pointer, we cast it to an NSWindow.
The objc crate wraps the objective-c runtime library, and basically just calls objc_msgSend which takes objective-c selectors and arguments. However it doesn't provide Cocoa or UIKit types or enum values, so we basically cast the window's opaque pointer to an NSObject, and call the setLevel selector. Also since we don't have the NSMainMenuWindowLevel, we have to find it from the Cocoa headers, which is 24. This allows the window to be above the level of the system menu bar. Macos however, doesn't allow decorated windows to stay on top of the menu bar, so we remove the decorations using set_border(false). The last call to resize just move the window back on top of the menu bar after setting its level.

**Reply by Lydanne** [reply]
Awesome, this is a really great project, I'll keep following it, thank you for your help.
